README
FusionCatcher
=============

Somatic fusion genes finder for RNA-seq data

Description
-----------
FusionCatcher searches for somatic novel/known fusion genes, translocations, and 
chimeras in RNA-seq data (paired-end reads from Illumina NGS platforms like 
Solexa/HiSeq/NextSeq/MiSeq) from diseased samples.

The aims of FusionCatcher are:
 * very good detection rate for finding candidate somatic fusion 
   genes (see somatic mutations; using a matched normal sample is 
   optional; several databases of known fusion genes found in healthy 
   samples are used as a list of known false positives; biological 
   knowledge is used, like for example gene fusion between a gene and 
   its pseudogene is filtered out),
 * very good RT-PCR validation rate of found candidate somatic fusion 
   genes (this is very important for us),
 * very easy to use (i.e. no a priori knowledge of bioinformatic 
   databases and bioinformatics is needed in order to run FusionCatcher BUT 
   Linux/Unix knowledge is needed; it allows a very high level of control 
   for expert users),
 * to be as automatic as possible (i.e. the FusionCatcher will choose 
   automatically the best parameters in order to find candidate somatic 
   fusion genes, e.g. finding automatically the adapters, quality trimming 
   of reads, building the exon-exon junctions automatically based on the 
   length of the reads given as input, etc. while giving also full control 
   to expert users) while providing the best possible detection rate for 
   finding somatic fusion genes (with a very low rate of false positives 
   but a very good precision).


Download/install/update FusionCatcher
-------------------------------------

Use this one-line command:

    wget http://sourceforge.net/projects/fusioncatcher/files/bootstrap.py -O bootstrap.py && python bootstrap.py --download


Manual
------
A detailed manual is at doc/MANUAL or [here](http://code.google.com/p/fusioncatcher/wiki/Manual).

Citing
------
D. Nicorici, M. Satalan, H. Edgren, S. Kangaspeska, A. Murumagi, O. Kallioniemi, 
S. Virtanen, O. Kilkku, FusionCatcher – a tool for finding somatic fusion genes 
in paired-end RNA-sequencing data, bioRxiv, Nov. 2014, [DOI:10.1101/011650](http://dx.doi.org/10.1101/011650)


NEWS
Version Release History for FusionCatcher
=========================================

Version 1.35 -- 30 June 2022
----------------------------
* add support for Ensembl version 106
* fixed minor bugs

Version 1.33 -- January 22, 2021
---------------------------------
 * increased sensitivity for TLX1, TLX2, and TLX3 fusions
 * added support for  using only the first N reads in case for very large FASTQ files
 * added a required dependency to FASTQTK tool: https://github.com/ndaniel/fastqtk
 * fixed several major/minor bugs
 * updated manual

Version 1.30 -- December 14, 2020
---------------------------------
 * added support for Ensembl version 102
 * added VCF format support for outputing the found candidate fusion genes
 * increased sensitivity for TLX3 fusions
 * fixed several major/minor bugs
 * updated manual

Version 1.20 -- November 8, 2019
--------------------------------
 * added support for Ensembl version 96, 97, and 98
 * increased sensitivity by adding an extra anchor
 * added --limitOutSJcollapsed for better control of STAR aligner
 * fixed memory bug leading to seg-fault in STAR aligner
 * improved handling of low entropy reads
 * size of built files for different organisms has been reduced
 * fixed several minor bugs
 * updated manual

Version 1.10 -- April 10, 2019
--------------------------------------
 * added support for Ensembl version 91, 92, 93, 94, and 95
 * fixed bug regarding randomness of mapping for Bowtie and STAR aligners
 * fixed bug regarding low complexity of fusion junction sequences
 * label for rt-circRNAs
 * fixed several minor bugs
 * updated manual


Version 1.00 - December 8, 2017
---------------------------------------
 * fixes for supporting Ensembl 90
 * increased sensitivity for IGH fusions
 * fixed bug regarding viral sequences when building new files for a new organism
 * fixed several minor bugs
 * updated manual


Version 0.99.7d beta - October 30, 2017
---------------------------------------
 * added support for Ensembl version 90
 * fixed several minor bugs
 * updated manual


Version 0.99.7c beta - June 30, 2017
---------------------------------------
 * added support for Ensembl versions 87, 88, and 89
 * added support for optionally forcing the execution of tools/scripts by using the corresponding absolute paths from the configuration file
 * added support for two versions of Bowtie aligner, that are 1.1.2 and 1.2
 * increased sensivity for certain types of IGH fusions and certain sizes of input reads
 * predicted full sequences of fusion transcripts are given for cases when the fusion is in-frame
 * candidate fusion genes are labelled now with exon-exon when the fusion junction is using known exon borders
 * fixed bug affecting very long reads (e.g. MiSeq reads)
 * fixed bug affecting single-end reads
 * fixed bug which affects in rare cases the prediction of fusion junction
 * fixed several minor bugs
 * updated manual


Version 0.99.7b beta - February 24, 2017
---------------------------------------
 * fixed a rare bug affecting 'bowtie --ff --tryhard'
 * fixed several minor bugs
 * updated manual


Version 0.99.7a beta - February 17, 2017
---------------------------------------
 * added support for Ensembl version 86
 * speed improvements
 * added support for more databases of known fusion genes
 * fixed a rare bug affecting split-fast.py
 * fixed bug affecting the PATHs for some of tools used by FusionCatcher
 * fixed bug regarding ZIP64
 * added fixes from https://github.com/yhoogstrate
 * fixed several minor bugs
 * updated manual


Version 0.99.6a beta - May 8, 2016
---------------------------------------
 * added support for Ensembl version 84
 * improved handling of pseudogenes
 * added support for Saccharomyces Cerevisiae organism
 * more tuning for the parameters for finding fusions which involve IGH loci
 * added to installation test for CRLF2-IGH@ and DUX4-IGH@ fusions
 * default aligners to be used can now be set also by configuration file 'fusioncatcher/etc/configuration.cfg'
 * fixed a bug regarding memory allocation when multiple threads are used
 * limiting the maximum memory to maximum of 26 GB used by sort command (from coreutils)
 * fixed several minor bugs
 * updated manual


Version 0.99.5a beta - January 15, 2016
---------------------------------------
 * added support for Ensembl version 83 and human genome GRCh38.p5
 * fixed a bug which affected the sensitivity of finding fusions which involve IGH loci
 * default number of threads/CPUs to be used can now be set also by configuration file 'fusioncatcher/etc/configuration.cfg'
 * added to the installation test the IGH@-MALT1 fusion
 * added support for STAR v2.5.0c aligner
 * fixed several minor bugs
 * updated manual

Version 0.99.4e beta - December 8, 2015
-----------------------------------------
 * added support for Ensembl version 82
 * increased sensitivity for fusion genes which involve IGH loci
 * added support for database of candidate gene fusions found in healthy human RNA-seq samples from 27 different tissues http://www.mcponline.org/content/13/2/397
 * added support for GTEx database (with help of FusionAnnotator)
 * fixed a bug which was introducing a large runtime-slowdown when candidate IGH translocations are found
 * fixed several minor bugs
 * updated manual


Version 0.99.4d beta - September 15, 2015
-----------------------------------------
 * added support for Ensembl version 81 and human genome GRCh38.p3/hg38.p3
 * fixed bug where input reads longer than 130 bp would produce very large number of 
   found fusions
 * increased sensitivity for fusion genes which involve IGH loci
 * fixed a bug affecting the building of index/build files for non-human organisms
 * fixed several minor bugs
 * updated manual


Version 0.99.4c beta - June 24, 2015
-------------------------------------
 * added support for Ensembl version 80
 * added support for fusion genes from paper: D. Robinson, Integrative Clinical 

get_gtf.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
It downloads the annotation in GTF foramt for a given organism from Ensembl.


Author: Daniel Nicorici, Daniel.Nicorici@gmail.com

Copyright (c) 2009-2014 Daniel Nicorici

This file is part of FusionCatcher.

FusionCatcher is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

FusionCatcher is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with FusionCatcher (see file 'COPYING.txt').  If not, see
<http://www.gnu.org/licenses/>.

By default, FusionCatcher is running BLAT aligner
<http://users.soe.ucsc.edu/~kent/src/> but it offers also the option to disable
all its scripts which make use of BLAT aligner if you choose explicitly to do so.
BLAT's license does not allow to be used for commercial activities. If BLAT
license does not allow to be used in your case then you may still use
FusionCatcher by forcing not use the BLAT aligner by specifying the option
'--skip-blat'. Fore more information regarding BLAT please see its license.

Please, note that FusionCatcher does not require BLAT in order to find
candidate fusion genes!

This file is not running/executing/using BLAT.
"""
#ftp.ensembl.org/pub/current_fasta/homo_sapiens/dna
#
#Files:
#
#Homo_sapiens.NCBI36.*.dna.chromosome.**.fa.gz
#
#where:
#* is 54 (ensembl version)
#** is chromosome name e.g. 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,X,Y,MT

#Hint: use ftplib and gzip

import os
import sys
import ftplib
import gzip
import socket
import optparse
import concatenate
import shutil


if __name__ == '__main__':

    #command line parsing

    usage = "%prog [options]"
    description = """It downloads the annotation in GTF format for a given organism from Ensembl."""
    version = "%prog 0.10 beta"

    parser = optparse.OptionParser(usage=usage,description=description,version=version)

    parser.add_option("--organism",
                      action = "store",
                      type = "string",
                      dest = "organism",
                      default = "homo_sapiens",
                      help="""The name of the organism for which the chromosomes are downloaded, e.g. homo_sapiens, mus_musculus, etc. Default is '%default'.""")

    parser.add_option("--output",
                      action="store",
                      type="string",
                      dest="output_directory",
                      default = '.',
                      help="""The output directory where the chromosomes are stored. Default is '%default'.""")

    parser.add_option("--server",
                      action="store",
                      type="string",
                      dest="server",
                      default = "ftp.ensembl.org",
                      help="""The Ensembl server from where the chromosomes are downloaded. Default is '%default'.""")

    parser.add_option("--server-path",
                      action="store",
                      type="string",
                      dest="server_path",
                      default = "/pub/current_gtf/", #ftp://ftp.ensembl.org/pub/release-75/gtf/
                      help="""The path of Ensembl server from where the data is downloaded. Default is '%default'.""")
                      
    parser.add_option("--filter-chrom",
                      action="store_true",
                      dest="filter",
                      default = False,
                      help="""All chromsomes which are not 1..99,X,Y,MT, and UN will be filtered out. Default is '%default'.""")


    #command line parsing
    (options, args) = parser.parse_args()

    #
    # validate options
    #
    if (
        (not options.output_directory)
        ):
        parser.print_help()
        sys.exit(1)

    #download file Homo_sapiens.GRCh37.56.gtf.gz
    try:
        ftp = ftplib.FTP(options.server)
        print ftp.login()
        ftp.cwd(options.server_path+options.organism.lower())

        list_files = ftp.nlst()

        list_files = [el for el in list_files if el.lower().startswith(options.organism.lower()) and el.lower().endswith('.gtf.gz') ]
        if len(list_files)!=1:
            print "Too many files or too few were found!"
            print list_files
            sys.exit(1)
        filename = list_files[0]


        fid=open(os.path.join(options.output_directory,filename),'wb')
        ftp.retrbinary("RETR " + filename, fid.write)
        fid.close()

        ftp.close()
    except ftplib.all_errors, e:
        print 'FTP Error = ' + str(e.code)
        sys.exit(1)
    except Exception, e:
        print "Error: Generic exception!",str(e)
        sys.exit(1)


    f=gzip.open(os.path.join(options.output_directory,filename), 'rb')
    file_content=f.read()
    f.close()
    temp_file = os.path.join(options.output_directory,'temp_organism.gtf')
    fod=open(temp_file,'wb')
    fod.write(file_content)
    fod.close()
    os.remove(os.path.join(options.output_directory,filename))

    #keep only the chromosomes 1, 2, 3, ... 99, MT,
    allchr = [str(i) for i in range(1,100)]
    chromosomes = set(allchr+['X','Y','MT','UN'])
    #
    if options.filter:
        data = [line for line in file(temp_file,'r').readlines() if line.startswith("#") or (line.split("\t",1)[0].upper() in chromosomes)]
        file(os.path.join(options.output_directory,"organism.gtf"),"w").writelines(data)
        os.remove(temp_file)
    else:
        os.rename(temp_file,os.path.join(options.output_directory,"organism.gtf"))

    #
